### ISSUE TITLE
PostCSS Custom Properties plugin hangs on cyclic dependencies declared in non-`:root` rules

### ISSUE THREAD

**Issue by reesscot** [csstools/postcss-plugins#1356]
### Reproduction link

https://stackblitz.com/edit/postcss-custom-properties-bug

### Bug description

When the PostCSS Custom Properties plugin is run on a CSS string with a large number of custom properties, the process completely hangs with no output.

Screenshot showing it unresponsive.
<img width="969" alt="CleanShot 2024-04-10 at 14 08 27@2x" src="https://github.com/csstools/postcss-plugins/assets/6165315/e356516b-7490-4743-869d-015b45352954">
 
When run with two less lines of CSS custom properties, it compiles correctly:
![CleanShot 2024-04-10 at 14 10 24](https://github.com/csstools/postcss-plugins/assets/6165315/09511296-9fe5-4e64-b686-22cfe097b098)

Related: https://github.com/aws-amplify/amplify-ui/issues/5143

### Actual Behavior

Node process hangs

### Expected Behavior

CSS should be output, as happens when a few custom property uses are removed from the CSS string.

### Can you reproduce it with `npx @csstools/csstools-cli <plugin-name> minimal-example.css`?

Yes

### `npx` Output

None, the process completely hangs.

### Extra config

None, using postcss-custom-properties by itself.

### What plugin are you experiencing this issue on?

PostCSS Custom Properties

### Plugin version

13.3.6

### What OS are you experiencing this on?

Windows

### Node Version

18.18.0

### Validations

- [X] Follow our [Code of Conduct](https://github.com/csstools/postcss-plugins/blob/master/CODE_OF_CONDUCT.md)
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

### Would you like to open a PR for this bug?

- [ ] I'm willing to open a PR

**Comment by romainmenke** [comment]
Hi @reesscot,

The stackblitz link doesn't work.
Do you have an alternate link?

**Comment by romainmenke** [comment]
Also, can you provide information as text, not as screencasts?
I can't copy/paste from images or video, so takes more of my time to debug this when I don't have all the info as text :)

**Comment by reesscot** [comment]
@romainmenke Sorry about that, I changed the url of the stackblitz and forgot to update it here. Updated above, but here's the link:
https://stackblitz.com/edit/postcss-custom-properties-bug

**Comment by romainmenke** [comment]
As far as I can tell you got circular references between this massive amount of props.

```css
:root,
[data-amplify-theme] {
	--amplify-components-input-color: var(--amplify-components-fieldcontrol-color);
}

/* .... */

.amplify-input {
	--amplify-components-fieldcontrol-color: var(--amplify-components-input-color);
}
```

We normally guard against this, and I don't know why it hangs instead of throwing errors.
Will investigate, but you can remove the circular references to get unblocked.

**Comment by romainmenke** [comment]
We don't guard against cyclic dependencies that are created in rules with selectors other than `:root`, `html`, ...

This is going to take some time to get sorted.
But you can also just clear up your cyclic dependencies between props.

--------

Also consider poking however you can internally to sponsor our open source work ;)

**Comment by reesscot** [comment]
I'm looking into why we are doing this, as it does seem odd and we may be able to get rid of these. However, according to the spec, it doesn't look like this is a cyclic dependency: https://www.w3.org/TR/css-variables-1/#cycles

> It is important to note that [custom properties](https://www.w3.org/TR/css-variables-1/#custom-property) resolve any [var()](https://www.w3.org/TR/css-variables-1/#funcdef-var) functions in their values at computed-value time, which occurs before the value is inherited. In general, cyclic dependencies occur only when multiple custom properties on the same element refer to each other; **custom properties defined on elements higher in the element tree can never cause a cyclic reference with properties defined on elements lower in the element tree.** (my emphasis)

**Comment by romainmenke** [comment]
Interesting!
That does offer an elegant way forwards.

I've prepared a fix for this : https://github.com/csstools/postcss-plugins/pull/1357

-----

personal opinion: I really wish people would stop using this plugin 😅
It doesn't make sense to use it as anything other than a polyfill for very old browsers and this is no longer needed.

**Comment by romainmenke** [comment]
@reesscot This has been resolved.
Thank you again for reporting this.

**Comment by reesscot** [comment]
@romainmenke Agreed that it doesn't really make sense to be polyfilling custom properties at this point given the level of browser support. I opened an [issue in the nuxt tailwindcss repo](https://github.com/nuxt-modules/tailwindcss/issues/835) if you want to give it a thumbs up.
